File: algorithms_7.0/warmup/h.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using ll = long long;
using namespace std;

bool search(vector<int>& arr, int& x) {
    for (int i = 0; i < arr.size(); ++i) {
        if (arr[i] == x) {
            return true;
        }
    }
    return false;
}

string add(vector<int>& arr, int& x) {
    if (search(arr, x)) {
        return "ALREADY";
    } else {
        arr.push_back(x);
        return "DONE";
    }
}

void print(vector<int>& arr) {
    for (int i = 0; i < arr.size(); ++i) {
        for (int j = 0; j < i; ++j) {
            cout << ".";
        }
        cout << arr[i] << "\n";
    }
}


int main() {
    string command;
    int value, x;

    vector<int> a;

    while (getline(cin, command)) {
        if (command.substr(0, 3) == "ADD") {
            value = stoi(command.substr(4));
            cout << add(a, value) << endl;
        } 
        else if (command.substr(0, 6) == "SEARCH") {
            value = stoi(command.substr(7));
            if (search(a, value)) {
                cout << "YES" << endl;
            } else {
                cout << "NO" << endl;
            }
        } 
        else if (command == "PRINTTREE") {
            print(a);
        }
    }

    return 0;
}

File: algorithms_7.0/warmup/h.py
import sys


class BST:
    def __init__(self):
        self.size = 1 << 22
        self.tree = [None] * self.size
    
    def add(self, v):
        ind = 1
        while ind < self.size:
            if self.tree[ind] is None:
                self.tree[ind] = v
                return "DONE"
            elif self.tree[ind] == v:
                return "ALREADY"
            elif v  < self.tree[ind]:
                ind = ind << 1
            else:
                ind = ind << 1 | 1
        return "ERROR"
    
    def search(self, v):
        ind = 1
        while ind < self.size:
            if self.tree[ind] is None:
                return "NO"
            elif self.tree[ind] == v:
                return "YES"
            elif v  < self.tree[ind]:
                ind = ind << 1
            else:
                ind = ind << 1 | 1
        return "NO"
    
    def printtree(self, ind=1, depth=0):
        if ind >= self.size or self.tree[ind] is None:
            return
        self.printtree(ind << 1, depth=depth+1)
        print('.'*depth + str(self.tree[ind]))
        self.printtree(ind << 1 | 1, depth=depth+1)


bst = BST()

for line in sys.stdin:
    line = line.strip()
    if line.startswith("ADD"):
        s, value = line.split()
        print(bst.add(int(value)))
    elif line.startswith("SEARCH"):
        s, value = line.split()
        print(bst.search(int(value)))
    elif line == "PRINTTREE":
        bst.printtree()